R/STMF.R

```diff
@@ -0,0 +1,125 @@
+#' Read Short-Term Mortality Fluctuations data from HMD and construct a `vital` object for use in other functions
+#'
+#' `read_stmf` reads weekly mortality data from the Short-term Mortality Fluctuations (STMF)
+#' series available in the Human Mortality Database (HMD) <https://www.mortality.org/Data/STMF>)
+#' and constructs a `vital` object suitable for use in other functions.
+#'
+#' @param CNTRY Directory abbreviation from the HMD. For instance, Australia = "AUS".
+#' @return `read_stmf_data` returns a `vital` object combining the downloaded data.
+#'
+#' @author Sixian Tang
+#' @examples
+#' \dontrun{
+#' norway <- read_stmf_data(CNTRY = "NOR")
+#' }
+#'
+#'
+#' @export
+read_stmf_data <- function(CNTRY) {
+
+  # check if country code is available
+  ctrylist    <- getSTMFcountries()
+
+  ctrylookup  <- ctrylist |>
+    dplyr::select(-"link")
+
+  # get CNTRY
+  if (missing(CNTRY)){
+    cat("\nCNTRY missing\n")
+    if (interactive()){
+      CNTRY <- select.list(choices = ctrylookup$CNTRY, multiple = FALSE, title = "Select Country Code")
+    } else {
+      stop("CNTRY should be one of these:\n",paste(ctrylookup$CNTRY, collapse = ",\n"))
+    }
+  }
+  if (!(CNTRY %in% ctrylookup$CNTRY)){
+    cat("\nCNTRY not found\n")
+    if (interactive()){
+      CNTRY <- select.list(choices = ctrylookup$CNTRY, multiple = FALSE, title = "Select Country Code")
+    } else {
+      stop("CNTRY should be one of these:\n",paste(ctrylookup$CNTRY, collapse = ",\n"))
+    }
+  }
+  stopifnot(length(CNTRY) == 1)
+
+  # read STMF data
+  url <- paste0("https://www.mortality.org/File/GetDocument/Public/STMF/Outputs/", CNTRY, "stmfout.csv")
+  data <- read.csv(url, header = TRUE, stringsAsFactors = FALSE, check.names = FALSE)
+
+  stmf_to_vital(data)
+}
+#'
+#' Get STMF country list
+getSTMFcountries <- function () {
+  xpath <- "/html/body/div[1]/div/div/div/div/div[2]/table/tbody"
+  html <- rvest::read_html("https://www.mortality.org/Data/STMF")
+  links <- rvest::html_attr(rvest::html_elements(rvest::html_element(html, xpath = xpath),
+                                   "a"), "href")
+  cntry_names <- rvest::html_text2(rvest::html_elements(rvest::html_element(html,
+                                                       xpath = xpath), "a"))
+  # Create a tibble with country names and links
+  tab_main <- dplyr::mutate(tsibble::tibble(Country = cntry_names, link = links),
+                     # Extract country codes from the links
+                     CNTRY = sub("/File/GetDocument/Public\\\\STMF\\\\Outputs\\\\(\\w+)stmfout\\.csv", "\\1", .data$link))
+  return(tab_main)

```

Comment:
Later you throw away the link column. It would be neater to remove it in this function before returning `tab_main`.